features/FxA-71-AMO-migration-support/README.md

```diff
@@ -0,0 +1,189 @@
+
+![](migration-indicator.png)
+
+### Permission prompt
+
+Before redirecting back to AMO,
+we should prompt the user for permission
+to share their display-name and/or avatar
+if they have that information in their profile:
+
+![](permission-prompt.png)
+
+If the user has not set a display name or avatar,
+then that item should not appear in the prompt.
+
+### Smart signin/signup chooser
+
+AMO will provide the email address of an existing user
+when redirecting to Firefox Accounts
+for the migration flow.
+We should show the sign-in page
+when there is an existing account for that email,
+and the sign-up page when there is not.
+
+The provided email address
+should be pre-filled in the form
+in both cases.
+
+
+## Implementation details
+
+### Migration indicator
+
+In order to trigger the additional migration indicator,
+AMO should include `migration=amo` as a query parameter
+when beginning the OAuth login flow.
+
+In fxa-content-server,
+we will generalize the existing handling
+of the `migration` query parameter
+from the sync relier model,
+so that all reliers can import and report
+the source of a migration.
+We will also change the migration messaging
+from its current sync-specific text
+to the more generic version in the UX mocks above.
+
+### Permission prompt
+
+Since AMO is a trusted relier,
+it will not receive a permissions prompt by default.
+We will adopt the `prompt=consent` query parameter
+from [OpenID Connect](http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest)
+to allow them to request it explicitly.
+
+AMO should include `prompt=consent` as a query parameter
+when beginning the OAuth login flow.
+This will trigger fxa-content-server to show the permission prompt
+regardless of the trusted status of the relier.
+
+When showing the permission prompt
+the content-server should do the following:
+
+* Check and normalize the requested scope:
+  * If it is "profile" then replace it with
+    "profile:uid profile:email profile:display_name profile:avatar".
+  * If it contains some subset of the above profile sub-scopes
+    then leave it as is.
+  * If it contains any other value then error out,
+    as we cannot yet prompt for consent for other scopes.
+* If the request includes `access_type=offline` in the query parameters
+  then error out,
+  as we cannot yet prompt for consent to offline access.
+* Query fxa-profile-server
+  to determine what profile data the user has available.
+  * If they have not set a display name,
+    remove "profile:display_name" from the requested scopes.
+  * If they have not set a profile picture,
+    remove "profile:avatar" from the requested scopes.
+* Display the permission prompt
+  as shown in the UX mocks above,
+  including all and only the items that are
+  still referenced by the requested scope.
+
+This process will ensure that
+the user is not prompted to share information
+that the relier has not requested,
+or that they have not made available in their profile.
```

Comment:
It could also be an opportunity to fill in this information. Currently users have to somehow find their accounts.firefox.com/settings to realize entering this info can be done.